### PR TITLE
Add a section on Testing patterns

### PR DESCRIPTION
**Pull request by pbrisbin**

### CODE REVIEW

File: haskell-best-practices.md [hunk 1]
```diff
@@ -473,6 +473,214 @@ increases boilerplate.
...
+  ) where
+```
+
+### Imports
+
+Each project should have a `TestImport` used for all specs. Optionally, a
+`TestImport.IO` version can exist for writing non-App specs in `IO`.
+
+```hs
+-- Bad
```
Comment (mjgpy3) [inline haskell-best-practices.md]: Please add comments pointing to exactly what's bad to make it obvious for me, the dumb reader.

File: haskell-best-practices.md [hunk 2]
```diff
@@ -473,6 +473,214 @@ increases boilerplate.
...
+shouldApproximate :: Double -> Double -> Expectation
+
+-- Good
+shouldApproximate :: HasCallStack => Double -> Double -> Expectation
+
+-- Best (see below)
+shouldApproximate
+  :: (HasCallStack, MonadIO m)
```
Comment (mjgpy3) [inline haskell-best-practices.md:562]: Nice suggestion

File: haskell-best-practices.md [hunk 3]
```diff
@@ -473,6 +473,214 @@ increases boilerplate.
...
+```hs
+-- Bad
+assertEqual "..." a b
+
+-- Good
+a `shouldBe` b
+```
+
+Except for functions that return values as part of their assertions, like
+`assertJust`, or `assertRight`.
```
Comment (mjgpy3) [inline haskell-best-practices.md:591]: Should we settle on the use of these things vs e.g. `Just thingy <- ...`?

Comment (mjgpy3) [reply haskell-best-practices.md:591]: We also have like `shouldBeJust` floating around. What should we do with those?

File: haskell-best-practices.md [hunk 4]
```diff
@@ -473,6 +473,214 @@ increases boilerplate.
...
+`MonadFail(fail)` as `expectationFailure`.
+
+For example, a test about a user in the DB can incorporate that assertion,
+
+```hs
+-- Meh
+user <- assertJust =<< getDB userId
+userFoo user `shouldBe` Foo
+
+-- Nice
```
Comment (mjgpy3) [inline haskell-best-practices.md:624]: Cool! Thanks for mentioning. Should we prefer one pattern or another?

File: haskell-best-practices.md [hunk 5]
```diff
@@ -473,6 +473,214 @@ increases boilerplate.
...
+Just user <- getDB userId
+userFoo user `shouldBe` Foo
+```
+
+#### Example Descriptions
+
+Somehow the "should" Railsism took hold in our code-base. There's no need for it
```
Comment (mjgpy3) [inline haskell-best-practices.md:631]: :raised_hands: love it, death to "should", long live "should"

File: haskell-best-practices.md [hunk 6]
```diff
@@ -473,6 +473,214 @@ increases boilerplate.
...
+
+Also, avoid the word "successfully". When expecting something to happen, that
+you expect it to happen "successfully" can be assumed.
+
+```hs
+-- Bad
+it "should successfully deduplicate the users"
+  $ withGraph
+  $ do
+    omgI'mAllTheWayOverHereNow
```
Comment (mjgpy3) [inline haskell-best-practices.md:642]: LOL

File: haskell-best-practices.md [hunk 7]
```diff
@@ -473,6 +473,214 @@ increases boilerplate.
...
+- Using `prop` under `withApp`
+
+  Pull it out.
+
+- Using `shouldX` with `prop`
+
+  Use a pure `Bool` expression with `prop`.
+
+If you do run into ambiguity, prefer Type Applications (over type signatures or
+annotations) to resolve it.
```
Comment (mjgpy3) [inline haskell-best-practices.md]: Do you mind linking to `TypeApplications` docs? I didn't know about this extension prior to starting and many folks I've interviewed lately don't know about it eiter.

**Review by mjgpy3 — APPROVED** [review]
LGTM, some comments take 'em or leave 'em. Great to see the specs getting some love.

File: haskell-best-practices.md (hunk 3 above)
Comment (pbrisbin) [reply haskell-best-practices.md:591]: `assertJust` and `Just x <-` are the same in that you get the `x`, but `shouldBeJust` is different as it's an expectation and doesn't return the `x`: ``shouldBeJust = (`shouldSatisfy` isJust) :: m ()``.

(If I'm wrong and `shouldBeJust` actually exists in a form that _does_ return the `x`, well then it should be renamed/replaced by `assertJust`.)

As for `assertJust` vs a pattern-match fail, I'm always down to pick a One True Way, but the usage feels about 50/50 in our code-bases. Dare you take a stand?

File: haskell-best-practices.md (hunk 4 above)
Comment (pbrisbin) [reply haskell-best-practices.md:624]: Answered in above thread, I think.

**Comment by pbrisbin** [comment]
@mjgpy3 check it now

**Pull request merged**